Automate lang links

rikivillalba opened the pull request:
In response to #6613

This is my proposal to automate link generation:

- The logic is in .translation_links.R in the vignettes directory, and it is sourced in each vignette
- The message "Translations of this document are available in the following languages" is translated in the same source (there is a switch which defaults to english)
- Currently the output format is   en | fr | es | ..., that is, not a list but all inline. 

Note that currently, in documents with a TOC (datatable-faq.Rmd) the link list is below the toc in the packaged vignettes. That is unchanged as the TOC is built automatically by commonmark.
Also, #6394 should be considered.

tdhock commented on vignettes/.translation_links.R:
```diff
@@ -0,0 +1,32 @@
+block = sprintf(
+  "%s: %s\n", 
+  switch(lang,
+    fr = "Des traductions de ce document sont disponibles dans les langues suivantes",
```
This method for defining translations is non-standard. I think this would be better as gettext(...)
or to simplify, just use English. Only list the other languages at the top of the English vignette. and then the other vignettes can just link back to English.

rikivillalba replied:
Yes, I think in gettext however I don't figure out how to get the string for the appropriate language to compile the vignettes in all languages, other than set `Sys.setLanguage()` in each call.
Another option is to hardcode the header and let the code generate only the links.
But I think your second option is ok.

rikivillalba commented:
I made the following changes:
- The script is now is a function to which the "translations available" message is an arg, so it is not harcoded but passed from the vignette text.
- The message only prints if there are any translations other than current language (as now french version of datatable-joins.Rmd is missing)
- renamed .translation_links.R to _translation_links.R so ls can see it (it is part of the vignettes after all)
- NOTE: **I manually merge** datatable-keys-fast-subset.Rmd, datatable-reference-semantics.Rmd and datatable-secondary-indices-and-auto-indexing.Rmd because #6617 modified lines adjacent to this branch. 

_translation_links.R must be bundled in the package so r CMD check can build the vignettes. However that source is 'knitted' and present in each .R file generated from .Rmd and installed in doc.

French vignettes are no modified at all. An equivalent to "from English *original* vignette" message linking back to english might be added to translated vignettes, or code similar to that of the english ones, it is up to the owner to decide.  In the script case it will add all available links excluding that of the current vignette language determined from the folder name. 

As part of spanish translators there is work in progres in https://github.com/cienciadedatos/traduccion-vignettes-datatable 
tks

MichaelChirico commented on vignettes/_translation_links.R:
```diff
@@ -0,0 +1,25 @@
+# build a link list of alternative languages (may be character(0))
+.write.translation.links <- function(fmt) {
+    url = "https://rdatatable.gitlab.io/data.table/articles"
+    path = dirname(knitr::current_input(TRUE))
```
is there a litedown equivalent (#6583)?

rikivillalba replied:
The only alternative I found was `litedown:::.env$input`, that stores the file name (the input path to `litedown::fuse()`). Tested to work building the package from . and from ..

MichaelChirico commented on vignettes/_translation_links.R:
```diff
@@ -0,0 +1,25 @@
+      pattern = glob2rx(knitr::current_input(FALSE))
+    )
+    transl_lang = ifelse(dirname(translation) == ".", "en", dirname(translation))
+    block = if (length(which(transl_lang != lang))) {
```
```suggestion
    block = if (!all(transl_lang == lang)) {
```

MichaelChirico commented on vignettes/_translation_links.R:
```diff
@@ -0,0 +1,25 @@
+# build a link list of alternative languages (may be character(0))
```
```suggestion
# build a link list of alternative languages (may be character(0))
# idea is to look like 'Other languages: en | fr | de'
```

Just commenting on expected output to facilitate reading/future refactoring

MichaelChirico commented on vignettes/_translation_links.R:
```diff
@@ -0,0 +1,25 @@
+          transl_lang[transl_lang != lang],
+          file.path(url, sub("(?i)\\.Rmd$", ".html", translation[transl_lang != lang]))
+      )))
+    } else ""
```
WDYT about leaving unlinked text in for the current language instead?

Available languages: [fr](http://google.com) | [es](http://google.com)

vs

Available languages: en | [fr](http://google.com) | [es](http://google.com)

MichaelChirico commented on vignettes/datatable-benchmarking.Rmd:
````diff
@@ -25,9 +25,9 @@ h2 {
-* [French](https://rdatatable.gitlab.io/data.table/articles/fr/datatable-benchmarking.html)
+```{r echo=FALSE, file='_translation_links.R'}
+```
+`r .write.translation.links("Translations of this document are available in: %s")`
````
IIUC the idea is the translation team will just translate this string themselves directly in the .Rmd file, is that right? (seems fine, just want to confirm my understanding)

rikivillalba replied:
Exactly.

MichaelChirico commented on vignettes/_translation_links.R:
we also have this helper script, let's keep a consistent naming style (personally, I like your snake case style here):

https://github.com/Rdatatable/data.table/blob/master/vignettes/.check.translations.R

Also note that this should probably be added to `.Rbuildignore`? Maybe just ignore all `vignettes/_.*` files?

rikivillalba replied:
Sadly, adding it to .Rbuildignore causes R check to fail. It checks the vignettes to be able to be built from the tarball bundle alone.

MichaelChirico replied:
oh got it. I guess that's why it must be prefixed with `_` to not be "hidden", which in turn is why there's no R CMD check failure.

I am only concerned about R CMD check passing, it is fine to include it.

rikivillalba commented:
This is the language-link printing function, with the propositions, adapted to litedown. Must be after #6583

```r
# build a link list of alternative languages (may be character(0))
# idea is to look like 'Other languages: en | fr | de'
.write.translation.links <- function(fmt) {
  url = "https://rdatatable.gitlab.io/data.table/articles"
  # TODO: better litedown alternative to knitr::current_input()
  # fuse() setwd the .Rmd dir before evaluating chunks
  input_file = basename(litedown:::.env$input)
  path = dirname(normalizePath(input_file, mustWork = TRUE))
  
  if (basename(path) == "vignettes") {
    lang = "en"
  } else {
    lang = basename(path)
    path = dirname(path)
  }
  translation = dir(path,
    recursive = TRUE,
    pattern = glob2rx(input_file)
  )
  transl_lang = ifelse(dirname(translation) == ".", "en", dirname(translation))
  block = if (!all(transl_lang == lang)) {
    sprintf(fmt, paste(collapse = " | ", ifelse(
      transl_lang != lang,
      sprintf("[%s](%s)", transl_lang,  file.path(url, sub("(?i)\\.Rmd$", ".html", translation))),
      transl_lang     # print no link, only language code.
    )))
  } else ""
  block
}
```

MichaelChirico commented:
Filed https://github.com/yihui/litedown/issues/67 requesting a public interface equivalent to `knitr::current_input()`

MichaelChirico commented on vignettes/_translation_links.R:
```diff
@@ -0,0 +1,26 @@
+    )
+    transl_lang = ifelse(dirname(translation) == ".", "en", dirname(translation))
+    block = if (!all(transl_lang == lang)) {
+      sprintf(
+        fmt, paste(collapse = " | ", sprintf(
+          "[%s](%s)",
+          transl_lang[transl_lang != lang],
+          file.path(url, sub("(?i)\\.Rmd$", ".html", translation[transl_lang != lang]))
+      )))
```
(untested) I think this will get the effect of including the current language unlinked

```suggestion
      linked_transl = sprintf("[%s](%s)", transl_lang, file.path(url, sub("(?i)\\.Rmd$", ".html", translation))
      linked_transl[transl_lang == lang] = lang
      sprintf(fmt, paste(linked_transl, collapse = " | "))
```

rikivillalba replied:
I tested your suggestion succesfully, this is what looks like (pkgdown build)
![image](https://github.com/user-attachments/assets/26b88b99-405c-4334-aaab-6cfc24a6197b)
![image](https://github.com/user-attachments/assets/788fa81a-6304-441f-a9d2-da65c05f9529)
![image](https://github.com/user-attachments/assets/418c2415-b898-499b-b93b-fec042516810)

Link's urls are absolute, as they are expected to be followed from vignettes in a path where translations are not available. I don't know whether it is possible to specify to generate relative paths when built on pkgdown but keep absolute paths otherwise. 

Also: if there is no translation available, no message is generate at all (which I think it's ok).

Note. Missing ) in line 18.
```suggestion
      linked_transl = sprintf("[%s](%s)", transl_lang, file.path(url, sub("(?i)\\.Rmd$", ".html", translation)))
      linked_transl[transl_lang == lang] = lang
      sprintf(fmt, paste(linked_transl, collapse = " | "))
```

aitap commented:
My apologies to everyone involved for pushing e7e966ea1cd9497318c007f3b51855e033647665 straight to master. I forgot to create a branch first.

Does this PR need any additional work, besides cleaning up the merge conflicts, before it can be merged?

rikivillalba commented:
> My apologies to everyone involved for pushing [e7e966e](https://github.com/Rdatatable/data.table/commit/e7e966ea1cd9497318c007f3b51855e033647665) straight to master. I forgot to create a branch first.
> 
> Does this PR need any additional work, besides cleaning up the merge conflicts, before it can be merged?

I think this pr is linked to #6583, and to the litedown feature akin to `knitr::current_input()`: `litedown::get_context("input")` not yet in CRAN.

MichaelChirico commented:
> I think this pr is linked to #6583, and to the litedown feature akin to `knitr::current_input()`: `litedown::get_context("input")` not yet in CRAN.

I don't think that's required to merge here. If you think this is ready in its current form, let's merge, we can mark the litedown translation part for follow-up.

rikivillalba commented:
> > I think this pr is linked to #6583, and to the litedown feature akin to `knitr::current_input()`: `litedown::get_context("input")` not yet in CRAN.
> 
> I don't think that's required to merge here. If you think this is ready in its current form, let's merge, we can mark the litedown translation part for follow-up.

Added the code sections to generate the translation links. this is the status:
- All vignettes currently on master are updated, I merged the most recent content and added the code to russian and french vignettes for ease of merging.
- Building the articles via `pkgdown::build_articles()` works as expected.
Pending:
- The message header for translation links *is not translated yet* (i'm handing off to translators). The english is `"Translations of this document are available in: %s"`. Using gettext here would imply changing locale before running each vignette, someting I think is not worth and I didn' tested.
Observations:
- Merging should probably be manually solved for the code printing the links being just aside the original links.

MichaelChirico reviewed (approved):
I think this is good to go :shipit:

We can iterate more later.
